varada_trino_manager/query_json_jstack.py

```diff
@@ -0,0 +1,85 @@
+from time import sleep
+from threading import Thread, Event
+
+
+def run_query(query: str, client: Trino) -> dict:
+    _, stats = client.execute(query=query)
+    logger.info(f'Query: {query} QueryId: {stats["queryId"]} '
+                f'Query execution time: {round(stats["elapsedTimeMillis"]*0.001, 3)} Seconds')
+    return {"queryId": stats["queryId"], "elapsedTime": round(stats["elapsedTimeMillis"]*0.001, 3)}
+
+
+def collect_jstack(wait: int, keep_running: Event):
+    collect_commands = [
+        "echo '#######################################    NEW JSTACK CAPTURE   #######################################' | sudo tee -a /tmp/jstacks/jstack.txt",
+        "sudo jps | awk '/TrinoServer/ {print $1}' | sudo tee -a /tmp/jstacks/server.pid",
+        "sudo jstack $(sudo cat /tmp/jstacks/server.pid) | sudo tee -a /tmp/jstacks/jstack.txt || true",
+    ]
+    while keep_running.is_set():
+        parallel_ssh_execute(command="\n".join(collect_commands))
+        sleep(wait)
+
+
+def run(user: str, con: Connection, jsonpath: Path, query: str, jstack_wait: int, dest_dir: str):
+    try:
+        with open(jsonpath) as fd:
+            queries = load(fd)
+    except Exception:
+        logger.exception(f'Failed reading {jsonpath}')
+        raise exceptions.Exit(code=1)
+
+    if query not in queries:
+        logger.error(f'Query {query} is not in {queries.keys()}')
+        raise exceptions.Exit(code=1)
+
+    dir_commands = [
+        "sudo rm -rf /tmp/jstacks",
+        "sudo rm -rf /tmp/jstacks.tar.gz",
+        "sudo mkdir /tmp/jstacks",
+    ]
+    parallel_ssh_execute(command="\n".join(dir_commands))
+
+    with Trino(con=con, username=user) as trino_client:
+        """
+        Start collecting jstack as Thread, then run query
+        Once query has completed - stop collection
+        """
+        logger.info(f"Start collecting jstacks, interval of {jstack_wait}Sec")
+        keep_collecting_jstack = Event()
+        keep_collecting_jstack.set()
+        collect = Thread(target=collect_jstack, args=(jstack_wait, keep_collecting_jstack))
+        collect.start()
+
+        logger.info(f'Running query {query}')
+        _, stats = trino_client.execute(query=queries[query])
+
+        logger.info("Query completed, stopping jstacks collection")
+        keep_collecting_jstack.clear()
+        collect.join()
+
+    # download jstack collection
+    logger.info(f"Downloading jstacks to {dest_dir}/")
+    tar_commands = [
+        "sudo tar -C /tmp/jstacks -zcf /tmp/jstacks.tar.gz .",
+        "sudo chmod 777 /tmp/jstacks.tar.gz",
+    ]
+    parallel_ssh_execute(command="\n".join(tar_commands))
+
+    dir_path = Paths.logs_path if dest_dir is None else dest_dir
+    parallel_download(
+        remote_file_path="/tmp/jstacks.tar.gz", local_dir_path=dir_path
+    )
+    logger.info(f'Getting query json for query_id {stats["queryId"]}, saving to {dest_dir}/')
+    with PrestoRest(con=con) as trino_rest:
```

Comment:
since it's all a part of the same flow it makes sense to add the to the same `with` block - 
```
with PrestoRest(con=con) as trino_rest, open(f'{dest_dir}/{stats["queryId"]}.json', 'w') as fd:
    data = trino_rest.query_json(query_id=stats["queryId"]).json()
    dump(data, fd, indent=2)
```